[Spring MVC] 조태현 미션 제출합니다.

whxogus215 opened the pull request:
이번 주 미션은 기존에 진행했던 스프링 기초와는 확실히 난이도 차이가 있던 것 같습니다!
시험기간 덕분에 기간이 많이 주어져서 처음 접해본 `인터셉터`, `ArgumentResolver` 그리고 스프링 MVC의 핵심 기술인 `DispatcherServlet`을 공부할 수 있었습니다. 또한 Closed 된 리뷰, 트랙 리더분들의 리뷰를 참고한 덕분에 미션을 마무리할 수 있었습니다.

## 적용한 내용
- 저는 `auth`라는 패키지를 만들어 로그인/로그아웃, 인증과 관련된 로직을 해당 패키지로 분리시켰습니다. 따라서 기존 `MemberController`에 있던 로그인 관련 로직들이 `auth` 패키지의 `AuthController`에 있으니 리뷰 시 참고하시면 될 것 같습니다!
- 저는 Enum을 주로 사용하는 편인데요! 이번 미션에서도 Enum을 활용하여 토큰을 생성하거나 권한을 조회하는 데에 필요한 값들을 상수로 담았습니다. (`AuthConfig`)
- 트랙 리더분의 리뷰를 참고하여 사용자가 요청한 Cookie에 담긴 토큰("token") 값을 추출하기 위해 `@CookieValue`라는 스프링 어노테이션을 활용하였습니다. 해당 부분은 `AuthController`의 `loginCheck` 메서드에서 확인할 수 있습니다.

종강 이후, 미션까지 진행하느라 고생 많으셨습니다. 코드 리뷰를 통해 많은 이야기를 나눠보면 좋을 것 같습니다 😄

whxogus215 committed:
feat: 로그인 기능 구현

- 이메일, 비밀번호 입력 후, 로그인 시 JWT 토큰이 담긴 쿠키 반환

whxogus215 committed:
1단계 완료

whxogus215 committed:
docs: 리드미 추가

whxogus215 committed:
refactor: dto 패키지 분리

whxogus215 committed:
refactor: AuthController 구현

- MemberController에 있던 로그인 관련 메서드를 이동

whxogus215 committed:
refactor: JwtToken 관련 로직 리팩터링

- Jwt 관련 로직을 담당하는 JwtTokenManager 객체 추가
- Jwt 시크릿 키는 외부 파일로 관리

whxogus215 committed:
feat: CookeUtil 객체 추가

- 쿠키에 들어가있는 토큰을 추출하는 로직을 유틸 클래스로 분리

whxogus215 committed:
feat: AuthConfig Enum 추가

- Token을 추출하기 위한 key 값인 token을 저장

whxogus215 committed:
feat: LoginMemberArgumentResolver 구현

- 사용자가 요청한 토큰을 통해 인증된 멤버를 생성하는 리졸버 추가

whxogus215 committed:
2단계 완료

- /reservations POST 리팩터링
- 요청에 이름이 있을 경우, 해당 이름으로 저장
- 요청에 이름이 없고, 쿠키에 토큰이 있을 경우, 해당 토큰에서 얻은 이름으로 저장

whxogus215 committed:
3단계 완료

whxogus215 committed:
refactor: 인증 관련 코드 리팩터링

- AuthController의 loginCheck에서 @CookieValue로 토큰 값 추출
- CookieUtil 및 JwtTokenManager에서 AuthConfig Enum 객체를 전달받도록 변경

whxogus215 committed:
docs: 리드미 줄바꿈 추가

mete0rfish commented on src/main/java/roomescape/auth/AuthController.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    @GetMapping("/login/check")
+    public ResponseEntity loginCheck(@CookieValue(value = "token") String token) {
```
`@CookieValue` 로  `HttpServletRequest`와 `CookieUtil`을 다루지 않아 정말 간편한 거 같아요👍

whxogus215 replied:
https://github.com/next-step/spring-basic-roomescape-playground/pull/25
트랙 리더님의 PR을 참고했는데, 확실히 더 간단하고 직관적인 방법인 것 같습니다 😄

mete0rfish commented on src/main/java/roomescape/auth/AuthConfig.java:
```diff
@@ -0,0 +1,19 @@
+package roomescape.auth;
+
+public enum AuthConfig {
+
+    TOKEN("token"),
+    ROLE("role"),
+    NAME("name"),
+    ADMIN("admin");
+
+    private String value;
+
+    AuthConfig(final String value) {
+        this.value = value;
+    }
+
+    public String getValue() {
+        return value;
+    }
+}
```
확실히 enum 클래스를 사용하여 상수로 다루니 간편하네요!
그런데 enum name이 AuthConfig여서 설정 파일과 헷갈리는 거 같아요!

whxogus215 replied:
`WebConfig` 클래스를 만드는 것과 혼동이 되어서 `Config`라는 이름을 붙인 것 같네요.
`AuthConstant`라고 지으면 좀 더 관련있는 네이밍이 될 것 같은데 리뷰어님은 어떻게 생각하시나요?

mete0rfish replied:
상수를 나타내는 부분에서 좋은 것 같습니다!

mete0rfish commented on src/main/java/roomescape/member/dto/MemberLoginRequest.java:
```diff
@@ -0,0 +1,7 @@
+package roomescape.member.dto;
+
+public record MemberLoginRequest(
```
로그인 시 사용하는 DTO 같아요! 문제가 있다는 건 아니라 제 의견을 나누면 좋을 것 같아 남깁니다
저는 코드를 작성하다 `MemberRequest`와 `MemberLoginRequest`의 필드가 많이 중복되기에 새로운 DTO로 만드는 것을 꺼려했습니다. 개발하다 보면 DTO가 많아지는 경우가 많잖아요.
그래서 `Inner Class`로 **Member와 관련된 Request를 한 곳에서 관리**하면 어떨까라고 생각했습니다.

제가 참고했던 블로그도 첨부해보면 좋을 것 같아 남깁니다!
[Inner Class로 DTO 관리](https://lackofwillpower.tistory.com/166)

whxogus215 replied:
DTO를 `Inner Class`로 관리했을 시 장단점이 잘 정리된 글 같습니다! 뭐든지 장단점이 존재하는 만큼 정답은 없는 것 같네요! 리뷰어님의 의견처럼 중복되는 필드가 있는 DTO가 지금처럼 2개가 아니라 10개가 넘어간다면 이름을 짓는 것도 곤란해질 것 같아요. 덕분에 DTO에 대해 고민해볼 수 있었습니다 😄

mete0rfish commented on src/main/java/roomescape/member/MemberService.java:
```diff
@@ -1,17 +1,33 @@
         return new MemberResponse(member.getId(), member.getName(), member.getEmail());
     }
+
+    public String findMember(MemberLoginRequest memberLoginRequest) {
```
로그인 시 request 정보들로 회원을 찾는 로직 같아요!
메서드 명에 `Member를 찾는데 사용한 변수들`을 `By~`로 나타내도 좋을 것 같아요!

whxogus215 replied:
리뷰를 반영하여 메서드명을 수정해본다면 `findByMemberEmailAndPassword`와 같이 작성해볼 수 있겠네요! 좀 더 이름이 직관적이지만 이 메서드를 사용하는 `Controller`는 멤버를 조회하기 위해 `Email`과 `Password`라는 존재를 알게되는건 아닐까라는 생각이 들었습니다. 반면 서비스 계층에서 `DAO`에게 호출하는 메서드명은 `findByEmailAndPassword`인데, 이는 서비스 계층이 DB와 접근하는 `DAO`에게 `Member를 찾기 위해 Email과 Password를 사용해줘`라는 메시지를 전달하는 것처럼 보여 좀 더 객체지향적이다라는 생각이 들었습니다.

mete0rfish replied:
그렇군요! 추가로 궁금한 점이 있습니다!
현재 `findMember()`에서 DAO의 `findByEmailAndPassword`을 사용하여 member를 조회하고 있어요.

그런데 만약 `findById` 같은 메서드로 member를 검색하는 비즈니스 로직이 필요할 경우가 생긴다고 할게요.
그러면 이때 어떻게 처리해야할지 생각해봤습니다.

그렇게 되면 **member를 찾는 비즈니스 로직이 두 개가** 되기에, **두 로직을 구분하는 메서드 명이 필요하다고** 생각이 들었어요.

whxogus215 replied:
member를 찾기 위해 **ID를 사용하거나 이메일/비밀번호를 사용하는 두 가지 방법**이 비즈니스 로직에 필요하다고 한다면 리뷰어님 말씀처럼 각각의 이름을 분리해서 사용하는게 맞을 것 같네요! 좋은 의견 감사합니다 😄

mete0rfish commented on src/main/java/roomescape/auth/JwtTokenManager.java:
```diff
@@ -0,0 +1,47 @@
+                .parseClaimsJws(token)
+                .getBody();
+        // 토큰에 담긴 "key"를 추출
+        return claims.get(key.getValue(), String.class);
```
AuthConfig를 통해 Value를 가져오니 이 메서드 하나로 커버가 가능하네요!
저도 enum 클래스를 사용해보려고 노력해야겠네요 😁

mete0rfish commented on src/main/java/roomescape/auth/AuthController.java:
```diff
@@ -0,0 +1,58 @@
+import roomescape.member.dto.MemberLoginRequest;
+
+@RestController
+public class AuthController {
```
로그인 관련 컨트롤러를 별개로 구분하니 보기 편한 거 같아요!
이번 패키지 구조가 `도메인형 구조`로 되어 있는 것 같아요.
auth패키지 안에 있는 Interceptor와 CookieUtil, ArgumentResolver 같은 경우 `global`로 사용되는 반면에
`AuthController`의 경우 Member에 관해 `local 적으로` 사용된다는 느낌을 받았습니다.
**물론 정답은 없지만 의견은 나누면 좋을 것 같아 남겨봅니다!**

whxogus215 replied:
리뷰어님의 의견 덕분에 패키지 분리에 있어서 좀 더 기준이 명확해진 것 같아요! 그러고 보니 `Interceptor`와 `ArgumentResolver`의 경우, `reservation`에서도 쓰이고 있어서 `global`이라는 곳에 분리하는 게 맞을 것 같아요. 수정해보겠습니다.

mete0rfish reviewed (commented):
고생하셨어요! 이번 미션이 간단하면서도 파야할 부분이 많았던 것 같아요
enum 클래스를 잘 사용하신 걸 보고 저도 배워가네요👍

whxogus215 committed:
refactor: AuthConfig Enum 이름 수정

- AuthConfig에서 AuthConstant로 이름 변경(리뷰 반영)

whxogus215 committed:
refactor: 패키지 분리

- 리뷰반영
- ArgumentResolver, Interceptor, Util 클래스는 전역적으로 사용되므로 global 패키지로 분리
- Constant, Config는 각각 constant, config 패키지로 분리

mete0rfish reviewed (commented):
https://github.com/next-step/spring-basic-roomescape-playground/pull/28#discussion_r1654075799

whxogus215 committed:
refactor: 자바 스트림 적용

- for/if문으로 인한 depth를 자바 스트림을 사용해 줄임.
